aya/src/maps/info.rs

```diff
@@ -0,0 +1,173 @@
+//! Metadata information about an eBPF map.
+
+use std::{
+    ffi::CString,
+    num::NonZeroU32,
+    os::fd::{AsFd as _, BorrowedFd},
+    path::Path,
+};
+
+use aya_obj::generated::{bpf_map_info, bpf_map_type};
+
+use super::{MapError, MapFd};
+use crate::{
+    sys::{
+        bpf_get_object, bpf_map_get_fd_by_id, bpf_map_get_info_by_fd, iter_map_ids, SyscallError,
+    },
+    util::bytes_of_bpf_name,
+    FEATURES,
+};
+
+/// Provides information about a loaded map, like name, id and size.
+#[doc(alias = "bpf_map_info")]
+#[derive(Debug)]
+pub struct MapInfo(pub(crate) bpf_map_info);
+
+impl MapInfo {
+    pub(crate) fn new_from_fd(fd: BorrowedFd<'_>) -> Result<Self, MapError> {
+        let info = bpf_map_get_info_by_fd(fd.as_fd())?;
+        Ok(Self(info))
+    }
+
+    /// Loads map info from a map ID.
+    ///
+    /// Uses kernel v4.13 features.
+    pub fn from_id(id: u32) -> Result<Self, MapError> {
+        bpf_map_get_fd_by_id(id)
+            .map_err(MapError::from)
+            .and_then(|fd| Self::new_from_fd(fd.as_fd()))
+    }
+
+    /// The map type as defined by the linux kernel enum [`bpf_map_type`].
+    ///
+    /// Introduced in kernel v4.13.
+    pub fn map_type(&self) -> bpf_map_type {
```

Comment:
we can't re-export generated types as part of the public API, or we can never
guarantee stability. I think we need our own EbpfMapType/EbpfProgramType this
issue has come up before.

Reply:
Could I get your thoughts on which would make the most sense for the enum name: 😅
- `EbpfProgramType`
- `EbpfProgram`
- `ProgramType`

I'm not sure if your feedback from https://github.com/aya-rs/aya/pull/959#discussion_r1716884645 applies here, since the `aya::programs::Program` enum is already being used to represent the variants of the actual program implementations.

Personally, I'm kinda leaning more towards `EbpfProgramType` or `ProgramType`, since I feel the `Type` suffix would give better distinction from the `aya::programs::Program` enum.

Reply:
ProgramType and MapType look great :)